### ISSUE TITLE
Document refs archival strategy for monorepo

### ISSUE THREAD

**Issue by sluongng** [monorepo-book/monorepo-book.github.io#6]
# Overview

Hi folks,

I want to contribute the following doc as it's what we have been using for years internally,
but I am not sure where to put the doc so I didn't create a PR for it.

Would appreciate some pointers and early reviews if possible 🙏 

# Doc contents

In a relatively large repo, many refs (branches or tags or system refs like refs/pull-requests/* or refs/merge-requests/*) could slow down repo in many ways:

- More refs to iterate through when using old git client (serving protocol v1)
- More objects that are reachable when doing git-gc or git-repack on serverside
- Initial git-clone will download all the refs by default

However, refs are often very useful to keep around:

- Some deployment / release strategies use git tags as bookmarks for the deployed commit
- Branches could also be used the same way.

A work around for this is to introduce a ref is:

1. Create a fork of the monorepo(origin) as a fork remote
2. Create a scheduled job which pull all the refs from the origin remote, then push them to the fork remote
3. Delete all the refs in the origin remote that is older than x months
4. (Optional) Delete all the refs in the fork remote that is older than y months (given y>x)

This way users often will only have to interact with a 'thinner' version of the monorepo with a lot less refs while all the refs still can be fetch/recovered from the fork remote repository. The thinner version will also have less objects and therefore can perform git-repack / git-gc faster, more regularly to help speed up git push / git pull operations.

Bonus: each hosting platform has different way to optimize how forks are stored and deduplicates. Refer to GHES(1) and Gitlab (2) for specific storage optimization to save on storage spaces.

(1): https://git-scm.com/docs/gitnamespaces
(2): https://docs.gitlab.com/ee/development/git_object_deduplication.html#pool-repositories

**Comment by sluongng** [comment]
after 5 mins of googling how GHES handling forks, I cannot find official docs so I decided to link git namespaces doc instead. Feel free to hit me up with the correct link.

**Comment by derrickstolee** [comment]
Just a side note: forks normally allow working across a trust boundary, such as in open source code. Private monorepos do not typically have such a trust boundary, so using forks in this way is a way to treat a performance issue, but does cause user issues, especially around sharing code that hasn't merged. The downsides of this approach should be included as a warning.

**Comment by sluongng** [comment]
> Just a side note: forks normally allow working across a trust boundary, such as in open source code. Private monorepos do not typically have such a trust boundary, so using forks in this way is a way to treat a performance issue, but does cause user issues, especially around sharing code that hasn't merged. The downsides of this approach should be included as a warning.

@derrickstolee thanks for the quick reply. 

Let me try to rephrase it to make sure that I get your point:

---

Note: Depends on the Hosting Service, a Fork repository might have a different ACL than the Origin Monorepo. This could potentially enable read access to source code to unintended groups or users (can read Fork repository while not having access to Origin Monorepo)

For example: Let's say the monorepo is located at `https://githost.org.com/group1/monorepo.git` with user A and B and C having access to the code in `group1` but only A and B have access to read `group1/monorepo.git`. Then having a fork at `https://githost.org.com/group1/monorepo-fork.git` might means that user C might have read access to content of `group/monorepo.git` through `group1/monorepo-fork.git` unintendedly.

This setup should be coupled with mechanism to ensure that the access control over the Fork would be the same (or more strict) compare to the Origin Monorepo. For access control specific configuration, please refer to hosting vendor documentations.

---

Does that seem right?

**Comment by derrickstolee** [comment]
> > Just a side note: forks normally allow working across a trust boundary, such as in open source code. Private monorepos do not typically have such a trust boundary, so using forks in this way is a way to treat a performance issue, but does cause user issues, especially around sharing code that hasn't merged. The downsides of this approach should be included as a warning.
> 
> @derrickstolee thanks for the quick reply.
> 
> Let me try to rephrase it to make sure that I get your point:
> 
> Note: Depends on the Hosting Service, a Fork repository might have a different ACL than the Origin Monorepo. This could potentially enable read access to source code to unintended groups or users (can read Fork repository while not having access to Origin Monorepo)
> 
> For example: Let's say the monorepo is located at `https://githost.org.com/group1/monorepo.git` with user A and B and C having access to the code in `group1` but only A and B have access to read `group1/monorepo.git`. Then having a fork at `https://githost.org.com/group1/monorepo-fork.git` might means that user C might have read access to content of `group/monorepo.git` through `group1/monorepo-fork.git` unintendedly.
> 
> This setup should be coupled with mechanism to ensure that the access control over the Fork would be the same (or more strict) compare to the Origin Monorepo. For access control specific configuration, please refer to hosting vendor documentations.
> 
> Does that seem right?

I think you misunderstand what my concern is. I believe that the hoster can lock down the read privileges of forks properly (for example: you can fork a private repo but can't make that fork public). In the world of public repos, people fork so they can _write_ to a copy of the repository that they can _already read_. In monorepos, the contributors typically have write permissions to the monorepo, since that is necessary to approve pull requests for their peers.

What I _am_ worried about is that if everyone has their own fork of the monorepo, then user A pushed their code to their fork and user B needs to add A's fork in order to fetch that data. If A and B were instead pushing to the monorepo, then B could just fetch from the origin as normal and checkout A's branch.

_Edit:_ Also, this doesn't change anything about the scale issues under the server's responsibility, since it needs to consider all of the forks' refs when doing maintenance activities. The forks only resolve the symptom of a large number of refs in the ref advertisement during a fetch and the number of refs stored in each client.

**Comment by sluongng** [comment]
> I think you misunderstand what my concern is. I believe that the hoster can lock down the read privileges of forks properly (for example: you can fork a private repo but can't make that fork public). In the world of public repos, people fork so they can _write_ to a copy of the repository that they can _already read_. In monorepos, the contributors typically have write permissions to the monorepo, since that is necessary to approve pull requests for their peers.
> 
> What I _am_ worried about is that if everyone has their own fork of the monorepo, then user A pushed their code to their fork and user B needs to add A's fork in order to fetch that data. If A and B were instead pushing to the monorepo, then B could just fetch from the origin as normal and checkout A's branch.

Ah, this is more clear now.

Yes, this strategy is NOT to be confused with encourage 'forking' the monorepo regularly. In fact, we explicitly disable forking internally for monorepo use case to help guide our users toward contribute to the repo directly instead of through a fork.

Forking mechanism was used in the proposal because it is where Hosting Vendor often equipped with storage deduplication and cleanup optimization. In some cases, it also helps speed up the refs archive process since you don't need to re-upload the objects which are shared between the 2 copies.

> _Edit:_ Also, this doesn't change anything about the scale issues under the server's responsibility, since it needs to consider all of the forks' refs when doing maintenance activities. The forks only resolve the symptom of a large number of refs in the ref advertisement during a fetch and the number of refs stored in each client.

Not particularly true.

Internally, we don't use a fork for refs archive. It's an entirely separate copy of the repo with no object shared between the 2. In fact, the archive copy is stored on a separate set of hosts with slower storage medium than our primary copy. This enabled 2 things:
- We have another backup of our monorepo on top for other backup mechanism that vendor has provided.
- The repack / gc on our primary copy has a lot less objects to deal with and thus run a lot faster. With gitlab current cleanup strategies, this enable our bitmap index and commit-graph to be updated more frequently than our fork.

The tradeoff here is that during the archive process, we have to re-download and re-push the refs and objects from 1 place to another. But as it's a background job, it has no visible user impacts (except for those who clone all the branches and tags, then their refs list get updated fairly frequently for deleted refs in remote origin).

**Comment by sluongng** [comment]
Another caveats: For monorepo with LFS, this means that you will also need to replicate the LFS storage from 1 copy of the repo to the other for both copies to be able to function individually.

If you decided to use the same LFS store between the 2 repos, be mindful that clean up mechanism in the origin repo may prune your LFS objects for they are no longer reachable from the post-archival remaining refs. In that case, you might need to either disable the Git LFS cleanup from both repositories as neither do contain a complete set of refs to sufficiently do a reachability checks for objects in Git LFS.

**Comment by vtbassmatt** [comment]
For GHES, if you create a fork, it behaves as @derrickstolee mentioned. You'd have to create a "clone and push" copy of the repo with no connection to its upstream.

The way I've tried to structure the book symptoms -> diagnosis -> solution starters. This feels like it dives a little deeper than I'd prefer. You have some good symptoms and diagnosis; if I were writing this section, I'd probably end with "find ways to eliminate unused refs -- perhaps those older than **x** months". Automating a mirror and retaining refs on it is out of scope for monorepo success.

**Comment by sluongng** [comment]
> For GHES, if you create a fork, it behaves as @derrickstolee mentioned. You'd have to create a "clone and push" copy of the repo with no connection to its upstream.

@vtbassmatt I can switch up the writing to recommend using a stand-alone non-fork copy, then adding a note that using a fork will help save disk space on server storage, but come with a set of trade-offs(i.e. repack/gc remains slow)

> The way I've tried to structure the book symptoms -> diagnosis -> solution starters. This feels like it dives a little deeper than I'd prefer. You have some good symptoms and diagnosis; if I were writing this section, I'd probably end with "find ways to eliminate unused refs -- perhaps those older than x months". Automating a mirror and retaining refs on it is out of scope for monorepo success.

Good suggestion.

Symptoms: slow to git-pull and git-clone, repo size getting very high

Diagnosis: 
- Run `git ls-remote origin | wc -l` on client side to count refs being adverised by remote
- Different vendors may expose metrics in term of distribution of git protocol usages(v0, v2 etc..)
- Run some `git for-each-ref` command on serverside to help identify *stale* refs
- Run a `git clone --mirror` on server to create a full copy of the current repo to test pruning refs + gc. Verify result with `git count-objects -v -H` or by measuring clone/pull time from client side.

Solutions:
- Simple starter: Prune old refs.
- Advance: Archive old refs to a different copy first, then prune (refs multi-tier storage)

Let me know what you think about this general outline/format. I can start a new markdown doc in a PR which will get linked / referenced-to from [fetches-too-slow.md](./docs/fetches-too-slow.md)

**Comment by vtbassmatt** [comment]
A new section to `fetches-too-slow.md` sounds great. The heading could be "Prune old refs".

I have some more concerns about the content. I don't want it to become a cookbook of precise commands to run, especially in this case where we're recommending deleting data. The book should introduce concepts and give pointers on how to 1) learn more and 2) speak intelligently with experts (or your vendor). The proposed content is very command-oriented, which I think is a mistake.

I very much realize that I'm changing goalposts on you (or rather, introducing goalposts where there were none before). I appreciate your patience as I'm just discovering some of these thoughts as a result of our conversations. I _do_ think that the idea of reducing ref count should be included in the book someplace.

**Comment by sluongng** [comment]
> I have some more concerns about the content. I don't want it to become a cookbook of precise commands to run, especially in this case where we're recommending deleting data. The book should introduce concepts and give pointers on how to 1) learn more and 2) speak intelligently with experts (or your vendor). The proposed content is very command-oriented, which I think is a mistake.

This sounds great.

However I do remember my experience when I was newly onboarded to manage our current monorepo and how lost I felt back then. This book will be targeting technical users, I think having some pseudo-code: not to exact implementation but some example commands, might be helpful. For example, a book like [Software Engineering at Google](https://abseil.io/resources/swe-book) would include many sample code snippets to demonstrate the concepts (see page 37/602 in the free pdf version).

What I would propose is, at the minimum, to have some small basic snippets that can be executed on the git client side. Then we can add footnotes / references section pointing to vendor specific docs on how server side should be maintained?

```
# Examples

# Show all refs 'origin' remote is advertising to client before actual data transfer
# See: https://git-scm.com/docs/git-ls-remote.html for details usages
> git ls-remote origin
```

> I very much realize that I'm changing goalposts on you (or rather, introducing goalposts where there were none before). I appreciate your patience as I'm just discovering some of these thoughts as a result of our conversations. I do think that the idea of reducing ref count should be included in the book someplace.

Yeah hashing out ideas in GH issue has been great :)

**Comment by vtbassmatt** [comment]
You raise a good point. I want to be careful to keep the book accessible, but also provide real value to those feeling lost.

Here's an idea -- much like we have the prose-oriented problems/solutions section today, perhaps there's a whole new section of recipes/snippets/specific commands. Then we can hyperlink from problems -> recipes. This avoids the other problem I have with deeply technical, "run these commands"-type of content: it's _ugly_ to try and read when lines of code or script are interspersed with prose.
